extensions/wikia/UserLogin/js/UserLoginFacebook.js

```diff
@@ -170,19 +170,28 @@
 				this.loginCallback();
 			} else {
 				this.bucky.timer.start('loggedInCallback');
-				var qString = new QueryString(),
-				// TODO: special page URL matching needs to be consolidated. @see UC-187
-					returnTo = (wgCanonicalSpecialPageName &&
-						(wgCanonicalSpecialPageName.match(/Userlogin|Userlogout|UserSignup/))) ?
-						wgMainPageTitle : null;
+				var returnUrl = new QueryString(),
+					returnTo = returnUrl.getVal('returnto');
 
+				// See if we were passed a returnto URL.  If not and we're on a special page we shouldn't
+				// stay on, set returnto to be the main page
 				if (returnTo) {
-					qString.setPath(wgArticlePath.replace('$1', returnTo));
+					returnUrl.removeVal('returnto');
+				} else if (
+					wgCanonicalSpecialPageName &&
+					wgCanonicalSpecialPageName.match(/Userlogin|Userlogout|UserSignup/)
```

Comment:
Per our discussion, let's use some version of the existing back end logic for this.